sam_view.c

```diff
@@ -405,47 +414,292 @@ int parse_aux_list(auxhash_t *h, char *optarg) {
     return 0;
 }
 
+#define error(...) { \
+   print_error_errno("view",__VA_ARGS__); \
+   exit(EXIT_FAILURE); \
```

Comment:
Unlike bcftools, it is very uncommon in samtools to call `exit` to abort subcommands. This makes a difference for wrappers such as pysam, so it would be good to follow the existing samtools convention: use `print_error` or `print_error_errno` directly and return an error indication from subroutines etc, and this `error` macro will be unnecessary.

Reply:
Can you provide more detail about what the wrappers expect? It makes a big difference for the code structure. The `view` command is written in a way that is very hard to refactor and this was an easy solution to that.

Reply:
pysam has a habit of [calling samtools' `main`](https://github.com/pysam-developers/pysam/blob/50f784ed/samtools/samtools.pysam.c#L66) from python.  This of course causes problems if samtools calls `exit` as you get dumped out of the python interpreter.  Arguably it would be safer to call `fork` and run samtools in a separate process, but I guess they may have their reasons for using the method that they do.

The `error()` macro is only used in four functions, so it wouldn't be too difficult to adjust them so that they clean up any memory used and return error codes instead.

Reply:
Pysam rewrites `exit` via `longjmp` so you are no longer dumped out of the interpreter, but `exit` does then cause leaks.

The usual convention in samtools is to return from `main` rather than call `exit`.

Reply:
I've added a commit that changes the error handling so it doesn't call `exit()`, which should make pysam happier.

While going through this I also spotted and fixed a few places where error checking had been missed, so it was a useful exercise in that regard.  The most important of these was checking the return value from the iterator, so that it's possible to detect read errors.